### DISCUSSION TITLE
how to create smooth surface blood vessel model

### DISCUSSION THREAD

**Discussion by gobenpark** [pyvista/pyvista#2777] [Q&A] [answered]
i use pyvista for my project  medical viwer 

i want to make a blood vessel model with 15 groups of x y z points by making a points for every 15 dicom view ( mri )

So, I composed and expressed the points and the triangular face of each point manually with polydata, but it is not smooth. 

is there any solution?  sorry for my short english

this is point with custom triangle faces
![혈관 모델링1](https://user-images.githubusercontent.com/17901887/173240328-a320a83d-0f01-48fa-ba06-559b4298b425.JPG)

this is only plot points

![혈관포인트](https://user-images.githubusercontent.com/17901887/173240348-efe3863c-d96b-48b9-86e7-379131c7903d.JPG)

i want to make smooth and model like blood vessel not like angled

**Comment by tkoyama010** [comment, marked as answer]
You may be can smooth your mesh using [tesselate filter](https://docs.pyvista.org/api/core/_autosummary/pyvista.UnstructuredGridFilters.tessellate.html).

**Reply by gobenpark** [reply]
thanks for your comment ill test that filter

**Reply by gobenpark** [reply]
is it possible PolyData to tesselate ?  @tkoyama010 

`mesh = pv.PolyData(verts,faces)`
`mesh.tesselate()`

Tessellate filter is not supported for PolyData objects.

**Reply by tkoyama010** [reply]
You can use [pyvista.PolyData.cast_to_unstructured_grid](https://docs.pyvista.org/api/core/_autosummary/pyvista.PolyData.cast_to_unstructured_grid.html#pyvista.PolyData.cast_to_unstructured_grid) to cast PolyData to UnstructuredGrid.

**Reply by akaszynski** [reply]
I'm working on a filter that will allow you to smooth your mesh using Taubin smoothing. I think it's the way to go. The procedure is to first subdivide and then smooth your mesh. See #2829.

2D Tessellation is for the creation of a surface from a point cloud, but once you have your surface you can smooth it with one of our smoothing filters. Laplacian smoothing is implemented in [pyvista.PolyData.smooth](https://docs.pyvista.org/api/core/_autosummary/pyvista.PolyData.smooth.html), but it will cause the blood vessels to shrink a bit. The upcoming filter in #2829 will maintain the volume of the blood vessel without loosing volume.

@JacobBumgarner, you might have something to say as well since you built an entire app around this.

**Reply by gobenpark** [reply]
change configure all faces to mesh.delaunay_3d() and merge unstructured_grid and same
![캡처](https://user-images.githubusercontent.com/17901887/174826509-262be89d-f121-4dfb-9b6e-b98a53b53c35.JPG)
 
but  smooth not work 

```python
surf = grid.extract_surface()
surf.smooth(n_iter=1000,feature_smoothing=False)
surf.plot()
```

and thanks for your comments @akaszynski  
i hope solve this

**Reply by akaszynski** [reply]
```py
surf.smooth(n_iter=1000,feature_smoothing=False)
```
That's not `inplace` so you would need to do:

```py
surf = grid.extract_surface()
surf = surf.smooth(n_iter=1000,feature_smoothing=False)
surf.plot()
```

But you'll get better results with:
```py
surf = grid.extract_surface().triangulate().subdivide(2)
surf = surf.smooth(n_iter=1000, feature_smoothing=False)
surf.plot()
```

**Reply by gobenpark** [reply]
oh thanks  

i edited code  like your recommend `surf = grid.extract_surface().triangulate().subdivide(2)`
and taubin filter #2829  like 
![캡처](https://user-images.githubusercontent.com/17901887/174834807-8ea808a7-7830-41ac-839f-6c308caaf63b.JPG)

```python
n_iter = 10000

mesh = pv.read("samplemesh.vtk")
mesh_lap = mesh.smooth(n_iter=n_iter, relaxation_factor=0.01)
smoothed = vtk.vtkWindowedSincPolyDataFilter()
smoothed.SetInputData(mesh)
smoothed.SetNumberOfIterations(1000000)
smoothed.SetPassBand(1)
smoothed.SetNormalizeCoordinates(False)
smoothed.Update()

p = pv.Plotter(shape=(1,2))
p.subplot(0, 0)
p.add_title("Laplace")
p.add_mesh(mesh, opacity=0.4, color="white")
p.add_mesh(mesh_lap, color="red")

p.subplot(0, 1)
p.add_title("Taubin")
# p.add_mesh(mesh, opacity=0.4, color="white")
p.add_mesh(smoothed.GetOutput(), color="white")
p.link_views()
p.show()
```

now more similar blood vessel

**Comment by gobenpark** [comment]
here my code and data.json (https://drive.google.com/file/d/1bVCSHRiSkfxyZUWxwmYBAP8Bl6vxHC2s/view?usp=sharing) 

but same apply tessellate() 

```python
import matplotlib.pyplot as plt
import pyvista as pv
from mpl_toolkits.mplot3d import Axes3D

fig = plt.figure()
ax = Axes3D(fig, auto_add_to_figure=False)
fig.add_axes(ax)

import json
from collections import defaultdict

def CreateFaces(l1:list,l2:list) -> list:
    result = []

    for i in range(-1,len(l1)-2,1):
        for j in range(0,len(l1)-1,1):
            for k in l2:
                result.append([l1[i],k,l1[j]])

    for i in range(-1,len(l2)-2,1):
        for j in range(0,len(l2)-1,1):
            for k in l1:
                result.append([l2[i],k,l2[j]])

    return result

if __name__ == '__main__':
    f = open('data.json')
    data = json.load(f)

    verts = []

    x = []
    y = []
    z = []
    for k, v in data.items():
        for i in v:
            x.append(i[0])
            y.append(i[1])
            z.append(int(k) * 10)

    verts = list(zip(x, y, z))

    indexs = defaultdict(list)

    for k, v in enumerate(verts):
        indexs[v[2]].append(k)

    total = []
    keys = indexs.keys()
    indexcopy = indexs.copy()

    for k in range(0,len(keys)-2,1):
        t1 = indexcopy[k*10].copy()

        t2 = indexcopy[(k + 1)*10].copy()

        for j in CreateFaces(t1, t2):

            total.append([3]+j)

    mesh = pv.PolyData(verts,total)
    grid = mesh.cast_to_unstructured_grid()
    tessellated = grid.tessellate()
    tessellated.clear_data()  # cleans up plot
    tessellated.plot(show_edges=False)

```
